### PR TITLE
Fixes 371

### PR DESCRIPTION
**Pull request by atareshawty**
Introduces the concept of `orginal_spot` versus and `current_spot`. When checking for viewable challenges, it will now check to see if the challenge spot is your row or if any challenges involve user's who's `original_row` is the same as a squad leader's

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -103,27 +105,28 @@ def can_member_challenge_for_performance?(performance)
 
   def performing_member_has_spot
     return if admin? || director?
-    return unless spot.nil?
-    errors.add(:spot, 'can\'t be blank')
+    return unless current_spot.nil?
+    errors.add(:current_spot, 'can\'t be blank')
```
Comment (kylekthompson) [inline app/models/user.rb]: How about 
```ruby
errors.add(:current_spot, 'can\'t be blank') if current_spot.nil?
errors.add(:original_spot, 'can\'t be blank') if original_spot.nil?
```

File: db/migrate/20170929220708_add_original_spot_to_user.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    # then their spot's never been changed
+    # If a user's spot has never been opened, but they've been in a challenge, then
+    # their first UserChallenge.spot is their original spot
+    # If user's spot has been opened, but they've never been in a challenge, there are two options
+    # - If the user isn't an alternate, then they haven't lost their spot yet, so use the current spot
+    # - If the user is an alternate, then they lost their spot. Their current spot was used as a UserChallenge.spot
+    # - in a challenge for the performance for which their spot was opened
+    # If the user's spot has been opened and they've been in a challenge, we'll have to do some more digging
+    # - If the user's first user_challenge was created before their first discipline action, then user the UserChallenge.spot
+    # - Else
```
Comment (kylekthompson) [inline db/migrate/20170929220708_add_original_spot_to_user.rb]: Missing comments for else

File: db/migrate/20170929220708_add_original_spot_to_user.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
+class AddOriginalSpotToUser < ActiveRecord::Migration[5.0]
+  def up
+    add_column :users, :original_spot_id, :integer, index: true
+    add_foreign_key :users, :spots, column: :original_spot_id
+
+    User.performers.includes(:current_spot, :user_challenges, :challenges, :discipline_actions).each do |user|
```
Comment (kylekthompson) [inline db/migrate/20170929220708_add_original_spot_to_user.rb]: Transaction

File: db/migrate/20170929220708_add_original_spot_to_user.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    # If the user's spot has been opened and they've been in a challenge, we'll have to do some more digging
+    # - If the user's first user_challenge was created before their first discipline action, then user the UserChallenge.spot
+    # - Else
+    if !has_discipline_action && !been_in_a_challenge
+      user.current_spot
+    elsif !has_discipline_action && been_in_a_challenge
+      user.user_challenges.order(id: :asc).first.spot
+    elsif has_discipline_action && !been_in_a_challenge && user.alternate?
+      performance = user.discipline_actions.join(:performances).where(open_spot: true).order(id: :asc).first.performance
+      challenge = Challenge.joins(:user_challenges).where('performance_id = ? AND user_challenges.spot_id = ?', performance.id, user.current_spot.id)
```
Comment (kylekthompson) [inline db/migrate/20170929220708_add_original_spot_to_user.rb]: Why the SQL here? 

`where(performance_id: performance.id, user_challenges: { spot_id: user.current_spot.id })` should handle that

File: db/migrate/20170929220708_add_original_spot_to_user.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    else
+      first_action = user.discipline_actions.where(open_spot: true).order(id: :asc).first
+      first_user_challenge = user.user_challenges.first
+      if first_user_challenge.created_at < first_action.created_at
+        first_user_challenge.spot
+      else
+        current_spot = user.current_spot
+        first_challenge_after_open_spot = first_user_challenge.challenge
+        performance_before_first_challenge = Performance.where('id < ?', first_challenge_after_open_spot.performance.id).order(id: :desc).first
+        challenge = Challenge.joins(:user_challenges).where('performance_id = ? AND user_challenges.spot_id = ?', performance_before_first_challenge.id, current_spot.id)
```
Comment (kylekthompson) [inline db/migrate/20170929220708_add_original_spot_to_user.rb]: Similar comment as above.

File: lib/sql/performance/challengeable_users.sql [hunk 6]
```diff
@@ -31,7 +31,7 @@ JOIN (
   ) disciplines_for_performance
   ON users.buck_id = disciplines_for_performance.user_buck_id
   WHERE users.instrument = %{instrument} AND users.part = %{part} AND NOT users.buck_id = '%{buck_id}'
-) users_disciplines ON s.id = users_disciplines.spot_id
+) users_disciplines ON s.id = users_disciplines.current_spot_id
```
Comment (kylekthompson) [inline lib/sql/performance/challengeable_users.sql]: Is this SQL still used anywhere?

File: db/migrate/20170929220708_add_original_spot_to_user.rb (hunk 2 above)
Comment (atareshawty) [reply db/migrate/20170929220708_add_original_spot_to_user.rb]: lol thanks

File: client/src/data/user/propTypes.js [hunk 7]
```diff
@@ -11,7 +11,7 @@ const propTypes = {
   lastName: PropTypes.string.isRequired,
   part: PropTypes.oneOf(Object.values(helpers.parts)).isRequired,
   role: PropTypes.oneOf(Object.values(helpers.roles)),
-  spot: PropTypes.shape(spotProps)
+  current_spot: PropTypes.shape(spotProps)
```
Comment (atareshawty) [inline client/src/data/user/propTypes.js]: This should be camelCase

**Comment by atareshawty** [comment]
Everything looks good on staging and the migrations have already been run on prod, so 🚀  :shipit:

**Pull request merged**